File: Algorithms/dynamic-programming/longest-common-subsequence.py
""" LCS algorithm in python: """


""" utility function to print array:"""

def printArray(array):
    for i in range(len(array)):
        for j in range(len(array[0])):
            print(array[i][j], end= " ")
        print()

""" function to find LCS: """
def lcsAlgo(S1, S2, m, n):
    L = [[0 for x in range(n + 1)] for x in range(m + 1)]

    """ building the matrix from the bottom way up: """

    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0 or j == 0:
                L[i][j] = 0
            elif S1[i - 1] == S2[j - 1]:
                L[i][j] = L[i - 1][j - 1] + 1
            else:
                L[i][j] = max(L[i][j - 1], L[i - 1][j])

    """index is the length of the longest common subsequence: """
    index = L[m][n]

    """print lcs array: """
    printArray(L)

    lcs = [""] * (index + 1)
    lcs[index] = ""

    i = m
    j = n 

    while i > 0 and j > 0:
        if S1[i - 1] == S2[j - 1]:
            lcs[index - 1] = S1[i - 1]
            i -= 1
            j -= 1
            index -= 1

        elif L[i - 1][j] > L[i][j - 1]:
            i -= 1
        else:
            j -= 1
    
    """ printing the subsequences: """

    print("S1 : " + S1 + "S2 : " + S2)
    print("LCS: " + "".join(lcs))


if __name__ == "__main__":
    S1 = "ACADB"
    S2 = "CBDA"
    m = len(S1)
    n = len(S2)
    lcsAlgo(S1, S2, m, n)

File: Algorithms/dynamic-programming/longest-common-subsequence.cpp
/* longest common subsequence in cpp: */

#include "iostream"
#include "cstring"

using namespace std;

void lcsAlgo(char * S1, char* S2, int m, int n){
    int lcsTable[m + 1][n + 1];
    /*building the matrix in a bottom up way:*/

    for (int i = 0; i <= m; i++){
        for (int j = 0; j <= n; j++){
            if(i == 0 || j == 0){
                lcsTable[i][j] = 0;
                }
            else if(S1[i - 1] == S2[j - 1]){
                lcsTable[i][j] = lcsTable[i - 1][j - 1] + 1;
                }
            else{
                lcsTable[i][j] = max(lcsTable[i - 1][j] , lcsTable[i][j - 1]);
            }
            }
        }
    int index = lcsTable[m][n];
    char lcsAlgo[index + 1];
    lcsAlgo[index] = '\0';

    int i = m, j = n;

    while(i > 0 && j > 0){
        if(S1[i - 1] == S2[j - 1]){
            lcsAlgo[index - 1] = S1[i - 1];
            i --;
            j --;
            index --;
        } else if(lcsTable[i - 1][j] > lcsTable[i][j - 1]){
            i --;
        } else{
            j --;
        }
    }

    /* printing the subsequence: */
    cout << " S1: " << S1 << endl << "S2: " << S2 << endl << " LCS: " << lcsAlgo << endl;

}


/* driver code: */

int main(){
    char S1[] = "ACADB";
    char S2[] = "CBDA";

    int m = strlen(S1);
    int n = strlen(S2);

    lcsAlgo(S1, S2, m, n);

}